testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketChannelNotYetConnectedTest.java

```diff
@@ -68,4 +78,68 @@ private static void checkThrowable(Throwable cause) throws Throwable {
             throw cause;
         }
     }
+
+    @Test
+    @Timeout(30)
+    public void readMustBePendingUntilChannelIsActive(TestInfo info) throws Throwable {
+        run(info, new Runner<Bootstrap>() {
+            @Override
+            public void run(Bootstrap bootstrap) throws Throwable {
+                System.out.println(bootstrap);
+                final ConcurrentLinkedQueue<String> queue = new ConcurrentLinkedQueue<String>();
+                NioEventLoopGroup group = new NioEventLoopGroup();
+                ServerBootstrap sb = new ServerBootstrap().group(group);
+                Channel serverChannel = sb.childHandler(new ChannelInboundHandlerAdapter() {
+                    @Override
+                    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                        queue.add("server channel active: " + ctx.channel());
+                        ctx.writeAndFlush(Unpooled.copyInt(42));
+                    }
+
+                    @Override
+                    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                        queue.add("server exception caught: " + cause.getMessage());
+                        cause.printStackTrace();
+                    }
+                }).channel(NioServerSocketChannel.class).bind(0).sync().channel();
+                queue.add("server bound");
+
+                final CountDownLatch readLatch = new CountDownLatch(1);
+                bootstrap.handler(new ChannelInboundHandlerAdapter() {
+                    @Override
+                    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+                        queue.add("client handler added: " + ctx.channel());
+                        assertFalse(ctx.channel().isActive());
+                        ctx.read();
+                    }
+
+                    @Override
+                    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                        queue.add("client read");
+                        ByteBuf buf = (ByteBuf) msg;
+                        assertThat(buf.readableBytes()).isEqualTo(Integer.BYTES);
```

Comment:
honestly I am not sure how likely this is but in theory you could read a `1 byte` ByteBuf only and then another 3 bytes. So maybe just sub-class `ByteToMessageDecoder` so we ensure to buffer everything ?